experimental/cuda2/pending_queue_actions.c

```diff
@@ -134,6 +140,73 @@ static void iree_hal_cuda2_queue_action_list_take_all(
   available_list->tail = NULL;
 }
 
+//===----------------------------------------------------------------------===//
+// Ready-list processing
+//===----------------------------------------------------------------------===//
+
+// Ready action atomic slist entry struct.
+typedef struct iree_hal_cuda2_atomic_slist_entry_t {
+  iree_hal_cuda2_queue_action_t* ready_list_head;
+  iree_atomic_slist_intrusive_ptr_t slist_next;
+} iree_hal_cuda2_atomic_slist_entry_t;
+
+// Ready action atomic slist.
+IREE_TYPED_ATOMIC_SLIST_WRAPPER(iree_hal_cuda2_ready_action,
+                                iree_hal_cuda2_atomic_slist_entry_t,
+                                offsetof(iree_hal_cuda2_atomic_slist_entry_t,
+                                         slist_next));
+
+// The ready-list processing worker's working/exiting state.
+typedef enum iree_hal_cuda2_worker_state_e {
+  IREE_HAL_CUDA_WORKER_STATE_IDLE_WAITING = 0,
+  IREE_HAL_CUDA_WORKER_STATE_WORKLOAD_PENDING = 1,
+  IREE_HAL_CUDA_WORKER_STATE_EXIT_REQUESTED = -1,
+  IREE_HAL_CUDA_WORKER_STATE_EXIT_COMMITTED = -2,
+} iree_hal_cuda2_worker_state_t;
+
+// The data structure needed by a ready-list processing worker thread to issue
+// ready actions to the GPU.
+//
+// This data structure is shared between the parent thread, which owns the
+// whole pending actions queue, and the worker thread; so proper synchronization
+// is needed to touch it from both sides.
+//
+// The parent thread should push a list of ready actions to ready_worklist,
+// update working_state, and give notification naccordingly.
+// The worker thread waits on the notification and checks working_state, and
+// pops from the ready_worklist to process. THe worker thread also monintors
+// exiting_state and stops processing if requested by the parent thread.
+typedef struct iree_hal_cuda2_working_area_t {
+  iree_notification_t notification;
+  iree_hal_cuda2_ready_action_slist_t ready_worklist;  // atomic
+  iree_atomic_int32_t working_state;                   // atomic
+  iree_atomic_int32_t exiting_state;                   // atomic
```

Comment:
reusing the same enum for two state fields is weird - either make an enum for both or combine into a single state field (best as less chance for things to get in an inconsistent state where there's work pending _and_ the worker has exited, etc).

Reply:
SG. Fixed to have one state. Agreed that it makes it more "atomic".